cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/monitor/MetricsMessageAggregator.java

```diff
@@ -0,0 +1,278 @@
+
+  MetricsMessageAggregator(SchemaGenerator schemaGenerator,
+      DatumWriterFactory writerFactory, DatumReaderFactory readerFactory,
+      long aggregationWindowSeconds) {
+    TypeToken<MetricValues> metricValueType = TypeToken.of(MetricValues.class);
+    try {
+      this.schema = schemaGenerator.generate(metricValueType.getType());
+    } catch (UnsupportedTypeException e) {
+      throw new RuntimeException("Failed to generate schema for MetricValues",
+          e);
+    }
+    reader = readerFactory.create(metricValueType, schema);
+    recordWriter = writerFactory.create(metricValueType, schema);
+    this.cachemap = new HashMap<>();
+    this.decodingInputStream = new ResettableByteArrayInputStream();
+    this.decoder = new StringCachingDecoder(
+        new BinaryDecoder(decodingInputStream), cachemap);
+    this.encodeOutputStream = new ByteArrayOutputStream(1024);
+    this.encoder = new BinaryEncoder(encodeOutputStream);
+    this.aggregationWindowSeconds = Math.max(1, aggregationWindowSeconds);
+  }
+
+  public Iterator<Message> aggregate(Iterator<Message> input) {
+    cachemap.clear();
+    if (!input.hasNext()) {
+      return input;
+    }
+
+    Map<MetricsMapKey, MetricsAggregator> aggregated = new HashMap<>();
+    String[] messageId = new String[1];
+    long currentTimestamp = System.currentTimeMillis() / 1000;
+    // MetricValues will be aggregated across windows of size "aggregationWindowSeconds".
+    // We align the current time to the end of the last widow. This is done since in most cases,
+    // it is expected that all the MetricValues were emitted very recently. The offset ensures
+    // the last window completely lies in the past and maximizes the chance that all metrics lie within it.
+    long offset = (currentTimestamp % aggregationWindowSeconds) + 1;
+
+    input.forEachRemaining(message -> {
+      // Decode the metrics.
+      // Keep track of the last message ID as TopicRelayers need it to fetch messages in the next iteration.
+      messageId[0] = message.getId();
+      decodingInputStream.reset(message.getPayload());
+      MetricValues values;
+      try {
+        values = reader.read(decoder, schema);
+      } catch (IOException e) {
+        LOG.warn("Failed to decode metric message.", e);
+        return;
+      }
+
+      // Aggregate the metrics.
+      long timestamp = values.getTimestamp();
+      MetricsAggregator metricsAggregator = aggregated.computeIfAbsent(
+          new MetricsMapKey(values.getTags(),
+              getRoundedTimestamp(values.getTimestamp(), offset)),
+          (t) -> new MetricsAggregator());
+      for (MetricValue metric : values.getMetrics()) {
+        metricsAggregator.addMetric(metric, timestamp);
+      }
+    });
+
+    return aggregated.entrySet().stream().map((entry) -> entry.getValue()
+        .getAggregateMetricValues(entry.getKey().getTags(),
+            entry.getKey().getRoundedTimestamp())).map((metricValues) -> {
```

Comment:
Remove the `()` around `metricValues`. Also, it is better to align all the `.map` at the beginning of the line.

```
    return aggregated.entrySet().stream()
        .map(entry -> entry.getValue().getAggregateMetricValues(entry.getKey().getTags(),
            entry.getKey().getRoundedTimestamp()))
        .map(metricValues -> {
              ....
         }).iterator();
```

Reply:
Done.